### PR TITLE
Find cell

### PR DESCRIPTION
**Pull request by jRoeltgen**
Only CPU version has been written/tested. Binary search to find cell indices in rect_grid.

### CODE REVIEW

File: zero/gkyl_rect_grid.h [hunk 1]
```diff
@@ -29,6 +29,19 @@ struct gkyl_rect_grid {
 void gkyl_rect_grid_init(struct gkyl_rect_grid *grid, int ndim,
   const double *lower, const double *upper, const int *cells);
 
+/**
```
Comment (manauref) [inline zero/gkyl_rect_grid.h:32]: @jRoeltgen fix typos in these comments

File: zero/gkyl_rect_grid_priv.h [hunk 2]
```diff
@@ -0,0 +1,7 @@
+#pragma once
```
Comment (manauref) [inline zero/gkyl_rect_grid_priv.h:1]: @jRoeltgen 
i don't think you need a private header just to declare the signature of private functions. Either just place those functions in the .c file before they get used, or put the functions themselves in the private header.

Comment (manauref) [reply zero/gkyl_rect_grid_priv.h:1]: although I understand why one may want to do it this way, we just haven't AFAIK, so let's stick to the current model for now.

File: zero/rect_grid.c [hunk 3]
```diff
@@ -22,6 +24,138 @@ gkyl_rect_grid_init(struct gkyl_rect_grid *grid, int ndim,
   }
 }
 
+GKYL_CU_DH
+void gkyl_rect_grid_find_cell(struct gkyl_rect_grid *grid, const double *point, bool pickLower, const int **knownIdx, int *cellIdx){
+  int nDim = grid->ndim;
+  int searchNum = 0;
+  int searchDim[GKYL_MAX_DIM], *idxOut;
```
Comment (manauref) [inline zero/rect_grid.c:31]: I think you are not using idxOut. Delete.

File: zero/rect_grid.c [hunk 4]
```diff
@@ -22,6 +24,138 @@ gkyl_rect_grid_init(struct gkyl_rect_grid *grid, int ndim,
   }
 }
 
+GKYL_CU_DH
+void gkyl_rect_grid_find_cell(struct gkyl_rect_grid *grid, const double *point, bool pickLower, const int **knownIdx, int *cellIdx){
```
Comment (manauref) [inline zero/rect_grid.c:28]: If I understand correctly, you are declaring a bunch of variables at the beginning of your function. That's a common style in Fortran, and some C programmers. However I (or maybe even we) prefer that you declare variables as needed. So please move them to right before they are used. The benefit of this latter style is that you keep the definition closer to where it is used, and don't require the reader to memorize the definition to later understand or check its (correct) use

File: zero/rect_grid.c [hunk 5]
```diff
@@ -22,6 +24,138 @@ gkyl_rect_grid_init(struct gkyl_rect_grid *grid, int ndim,
...
+  int searchDim[GKYL_MAX_DIM], *idxOut;
+  int *dimTrans[GKYL_MAX_DIM];
+  int plusminus[2] = {-1,1};
+  bool allLessEq = true;
+  int iStart[GKYL_MAX_DIM], iEnd[GKYL_MAX_DIM], iMid[GKYL_MAX_DIM], iNew[GKYL_MAX_DIM];
+  int *cells, lowHighCellIdx[2*GKYL_MAX_DIM];
+  double lowerInDir[nDim], upperInDir[nDim];
+  double low, high;
+  
+  for(int d=0; d<nDim; d++){
```
Comment (manauref) [inline zero/rect_grid.c:40]: here and elsewhere, please use spacing like other gkeyll code.
In this case leave a space after **for**. Below leave a space after **if**, and before and after **else**.

File: zero/rect_grid.c [hunk 6]
```diff
@@ -22,6 +24,138 @@ gkyl_rect_grid_init(struct gkyl_rect_grid *grid, int ndim,
...
+    iEnd[d] = cells[searchDim[d]];
+    iMid[d] = 0;
+    iNew[d] = 0;
+  }
+  /* Below we use a binary search. That is, if the i-th coordinate of the point in
+   * question is below(above) the ith-coordinate of the current mid-point, we search
+   * the lower(upper) half along that direction in the next iteration.
+   */
+  while (allLessEq){
+    for(int d=0; d<searchNum; d++){
```
Comment (manauref) [inline zero/rect_grid.c:67]: Here and elsewhere:
- please remove the curly brackets in for loops that contain a single instruction.
- Leave at least 2 spaces between code and comment.
- Leave 1 space after // in your comment.

File: zero/rect_grid.c [hunk 7]
```diff
@@ -22,6 +24,138 @@ gkyl_rect_grid_init(struct gkyl_rect_grid *grid, int ndim,
...
+    if(isInCell(grid, point, iMid, dimTrans, knownIdx)){
+      //Check if neighboring cells also contain this point.
+      for(int k=0; k<searchNum; k++){
+	iNew[k] = iMid[k];
+	lowHighCellIdx[searchDim[k]] = iMid[k];
+	lowHighCellIdx[nDim+searchDim[k]] = iMid[k];
+      }
+      for(int i=0; i<searchNum; i++){	
+	for(int j=0; j<2; j++){
+	  iNew[i] = fmax(fmin(iMid[i] + plusminus[j],cells[i]),0);
```
Comment (manauref) [inline zero/rect_grid.c:80]: i think fmax and fmin are for floats/doubles (maybe I'm wrong).
There's a function like GKYL_MAX or something that you can use instead (see zero/gkyl_util.h).

File: zero/rect_grid.c [hunk 8]
```diff
@@ -22,6 +24,138 @@ gkyl_rect_grid_init(struct gkyl_rect_grid *grid, int ndim,
...
+  return inCell;
+}
+
+GKYL_CU_DH
+void InDir(const struct gkyl_rect_grid *grid, int *iIn, int *dimTrans[1], const int **knownIdx, double lowerInDir[], double upperInDir[]){
+  int nDim, checkIdx;
+  const double *dx, *lower;
+  nDim = grid -> ndim;
+  double testarr[nDim];
+  setbuf(stdout, NULL);
```
Comment (manauref) [inline zero/rect_grid.c:147]: what is this?

File: zero/rect_grid.c [hunk 9]
```diff
@@ -22,6 +24,138 @@ gkyl_rect_grid_init(struct gkyl_rect_grid *grid, int ndim,
...
+	allLessEq = false;
+	break;
+      }
+    }
+  }
+
+}
+
+GKYL_CU_DH
+bool isInCell(const struct gkyl_rect_grid *grid, const double *pIn, int *iIn, int *dimTrans[1], const int **knownIdx){
```
Comment (manauref) [inline zero/rect_grid.c:119]: please add some (very brief) comments describing this function and its inputs/outputs.

File: zero/rect_grid.c [hunk 10]
```diff
@@ -22,6 +24,138 @@ gkyl_rect_grid_init(struct gkyl_rect_grid *grid, int ndim,
...
+    if(lowerInDir[d]-eps>pIn[d] || upperInDir[d]+eps<pIn[d]){
+      inCell = false;
+      break;
+    }
+  }
+  return inCell;
+}
+
+GKYL_CU_DH
+void InDir(const struct gkyl_rect_grid *grid, int *iIn, int *dimTrans[1], const int **knownIdx, double lowerInDir[], double upperInDir[]){
```
Comment (manauref) [inline zero/rect_grid.c:142]: please add some (very brief) comments describing this function and its inputs/outputs.

File: zero/rect_grid.c (hunk 4 above)
Comment (manauref) [inline zero/rect_grid.c:28]: Biggest change I would recommend is to change knownIdx from a ** to a *. The latter will be easier to work with. We can assume that entries in knownIdx <0 are not known (like nil in the Lua code), and entries >=0 are known indices.
If you agree, please change throughout.

File: unit/ctest_rect_grid.c [hunk 11]
```diff
@@ -35,6 +35,139 @@ void test_grid_2d()
     }
 }
 
+void test_find_cell_1x(){
```
Comment (manauref) [inline unit/ctest_rect_grid.c:38]: please add some (very brief) comments to your testing functions to at least say which case you are testing in each bit of code. So future readers know the purpose of each bit.

**Comment by manauref** [comment]
@jRoeltgen you technically haven't tested your method on GPUs. Although you may have run on/with a GPU, it executed the code on the CPU.

To test it on the GPU you'll have to add code to uni/ctest_rect_grid_cu_.cu

**Comment by ammarhakim** [comment]
Please do not use camelCase for names. Please be consistent with the rest of the code and use underscores to name functions. Please make these changes before we consider the PR.

File: zero/rect_grid.c (hunk 4 above)
Comment (ammarhakim) [reply zero/rect_grid.c:28]: Do not use Fortran style :) I agree with Mana

File: zero/rect_grid.c (hunk 6 above)
Comment (jRoeltgen) [reply zero/rect_grid.c:67]: Should the curly brackets be neglected from the outer for loop in double for loops even when the inner for loop has multiple instructions (assuming the outer for loop only contains the inner for loop), such as the following?

for (int i=1; i<a; i++) {  // this curly bracket
   for (int j=1; j<b; j++) {
      x=i+j;
      .
      .
      .
   }
}

Comment (manauref) [reply zero/rect_grid.c:67]: I think not, good catch. An exception to my first bullet above is when that single instruction is another loop or switch

**Pull request closed without merging**